File: app/lib/features/settings/data/exchange_key.dart
/// Model for exchange API keys.
class ExchangeKey {
  final String id;
  final String exchange;
  final String label;
  final bool isActive;
  final String status; // untested, valid, invalid
  final String? errorMessage;
  final double? balance;
  final String? balanceUpdatedAt;
  final String? lastCheckedAt;
  final String? expiresAt;
  final String createdAt;
  final String updatedAt;

  ExchangeKey({
    required this.id,
    required this.exchange,
    required this.label,
    required this.isActive,
    required this.status,
    this.errorMessage,
    this.balance,
    this.balanceUpdatedAt,
    this.lastCheckedAt,
    this.expiresAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ExchangeKey.fromJson(Map<String, dynamic> json) => ExchangeKey(
        id: json['id'] as String,
        exchange: json['exchange'] as String,
        label: json['label'] as String? ?? '',
        isActive: json['is_active'] as bool? ?? true,
        status: json['status'] as String? ?? 'untested',
        errorMessage: json['error_message'] as String?,
        balance: (json['balance'] as num?)?.toDouble(),
        balanceUpdatedAt: json['balance_updated_at'] as String?,
        lastCheckedAt: json['last_checked_at'] as String?,
        expiresAt: json['expires_at'] as String?,
        createdAt: json['created_at'] as String,
        updatedAt: json['updated_at'] as String,
      );

  /// Human-readable status label.
  String get statusLabel {
    switch (status) {
      case 'valid':
        return '✅ Валидный';
      case 'invalid':
        return '❌ Невалидный';
      default:
        return '⏳ Не проверен';
    }
  }

  /// Days until expiration (null if no expiration set).
  int? get daysUntilExpiry {
    if (expiresAt == null) return null;
    final expires = DateTime.tryParse(expiresAt!);
    if (expires == null) return null;
    return expires.difference(DateTime.now()).inDays;
  }

  /// Human-readable expiration label.
  String? get expiryLabel {
    final days = daysUntilExpiry;
    if (days == null) return null;
    if (days < 0) return '⚠️ Истёк';
    if (days == 0) return '⚠️ Истекает сегодня';
    if (days <= 5) return '⏰ Осталось $days дн.';
    return '✅ $days дн.';
  }
}

File: backend/app/models/exchange_key.py
"""SQLAlchemy model for exchange API keys (encrypted at rest)."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, Float, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.core.database import Base


class ExchangeKey(Base):
    """Encrypted API key for a cryptocurrency exchange.

    Keys are encrypted at rest using Fernet symmetric encryption.
    """

    __tablename__ = "exchange_keys"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    exchange = Column(String(20), nullable=False)  # binance, mexc, bybit
    label = Column(String(100), default="")  # user-friendly name

    # Encrypted at rest
    api_key_encrypted = Column(Text, nullable=False)
    api_secret_encrypted = Column(Text, nullable=False)
    passphrase = Column(Text, nullable=True)  # For exchanges that need it (e.g. OKX)

    # Status
    is_active = Column(Boolean, default=True, nullable=False)
    status = Column(
        String(20), default="untested", nullable=False
    )  # untested, valid, invalid

    # Balance snapshot
    balance = Column(Float, nullable=True)
    balance_updated_at = Column(DateTime(timezone=True), nullable=True)

    # Metadata
    created_at = Column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
    last_checked_at = Column(DateTime(timezone=True), nullable=True)

    # Relationships
    user = relationship("User", backref="exchange_keys")

    def __repr__(self) -> str:
        return f"<ExchangeKey id={self.id} exchange={self.exchange} status={self.status}>"
